### PR TITLE
feat: update item doctype with new bloomtrace architecture

### PR DESCRIPTION
**Pull request by harshmule-epi**
> **Update Item doctype with new bloomtrace architecture**

### CODE REVIEW

File: erpnext/stock/doctype/item/item.js [hunk 1]
```diff
@@ -180,12 +184,35 @@ frappe.ui.form.on("Item", {
 				}
 			});
 		}
+		if (frm.doc.metrc_item_category) {
```
Comment (rvpasari) [inline erpnext/stock/doctype/item/item.js:188]: add a comment specify that this toggle is for fields that require strain name

File: erpnext/stock/doctype/item/item.py [hunk 2]
```diff
@@ -898,6 +905,77 @@ def validate_auto_reorder_enabled_in_stock_settings(self):
...
+			now = datetime.now()
+			request_data = {
+				"site_url": "manufacturing.bloomstack.io",
+				"customer_name": "Bloomstack",
+				"company_name": "Bloomstack India",
+				"license_number": "A12-0000015-LIC",
+				"rest_method": "POST",
+				"environment": "sandbox",
+				"doctype": "Item",
+				"doctype_data": {
```
Comment (rvpasari) [inline erpnext/stock/doctype/item/item.py]: dict(self)

File: erpnext/stock/doctype/item/item.py [hunk 3]
```diff
@@ -898,6 +905,77 @@ def validate_auto_reorder_enabled_in_stock_settings(self):
...
+					"parentfield": None,
+					"parenttype": None,
+					"idx": 0,
+					"docstatus": 0,
+					"company": "Bloom91"
+				}
+			}
+
+			# for item_category which requires strain
+			if self.strain_name:
```
Comment (rvpasari) [inline erpnext/stock/doctype/item/item.py:936]: this is fine

File: erpnext/stock/doctype/item/item.py [hunk 4]
```diff
@@ -898,6 +905,77 @@ def validate_auto_reorder_enabled_in_stock_settings(self):
...
+				frappe.msgprint(response.get('message'))
+			else:
+				response_message = {
+					400: "An error has occurred while executing request for METRC",
+					401: "Invalid or no authentication provided for METRC",
+					403: "The authenticated user does not have access to the requested resource for METRC",
+					404: "The requested resource could not be found (incorrect or invalid URI) for METRC",
+					429: "The limit of API calls allowed has been exceeded for METRC. Please pace the usage rate of API more apart",
+					500: "METRC Internal server error"
+				}
```
Comment (rvpasari) [inline erpnext/stock/doctype/item/item.py]: error messages should be for BloomTrace, not METRC

File: erpnext/stock/doctype/item/item.py [hunk 5]
```diff
@@ -898,6 +905,77 @@ def validate_auto_reorder_enabled_in_stock_settings(self):
...
+
+			# for item_category which requires strain
+			if self.strain_name:
+				request_data['doctype_data']['strain'] = self.strain_name
+				
+			# getting the item category from the Compliance Item Category for quantity type and mandatory unit
+			item_category = frappe.db.get("Compliance Item Category", self.metrc_item_category, ['quantity_type', 'mandatory_unit'])
+
+			# setting the unit uom and unit value as per the countbased quantity type
+			if item_category.get('quantity_type') == "CountBased":
```
Comment (rvpasari) [inline erpnext/stock/doctype/item/item.py:943]: any other quantity types that need to be handled?

**Review by rvpasari — CHANGES_REQUESTED** [review]

File: erpnext/stock/doctype/item/item.py (hunk 5 above)
Comment (harshmule-epi) [reply erpnext/stock/doctype/item/item.py:943]: No, there are 3 quatity types but for unit value and unit uom countbased quantity is the required quantity, and there is no other senarios for any other category.

File: erpnext/stock/doctype/item/item.py (hunk 4 above)
Comment (harshmule-epi) [reply erpnext/stock/doctype/item/item.py]: will, change that

File: erpnext/stock/doctype/item/item.py (hunk 2 above)
Comment (harshmule-epi) [reply erpnext/stock/doctype/item/item.py]: will check

File: erpnext/stock/doctype/item/item.js (hunk 1 above)
Comment (harshmule-epi) [reply erpnext/stock/doctype/item/item.js:188]: sure

File: erpnext/stock/doctype/item/item.py [hunk 6]
```diff
@@ -898,6 +905,72 @@ def validate_auto_reorder_enabled_in_stock_settings(self):
...
+				"doctype_data": data
+			}
+
+			# to map the metrc_item_name, metrc_uom, metrc_item_category
+			request_data['doctype_data']['name'] = self.metrc_item_name
+			request_data['doctype_data']['uom'] = self.metrc_uom
+			request_data['doctype_data']['item_category'] = self.metrc_item_category
+
+			# passing the hard coded modified_by
+			request_data['doctype_data']['modified_by'] = "neil@bloomstack.com"
```
Comment (rvpasari) [inline erpnext/stock/doctype/item/item.py]: this cannot be hard coded

File: erpnext/stock/doctype/item/item.py [hunk 7]
```diff
@@ -898,6 +905,72 @@ def validate_auto_reorder_enabled_in_stock_settings(self):
...
+			if item_category.get('quantity_type') == "CountBased":
+				if item_category.get('mandatory_unit') == "Volume": 
+					request_data['doctype_data']['unit_volume'] = self.metrc_unit_value
+					request_data['doctype_data']['unit_volume_uom'] = self.metrc_unit_uom
+				if item_category.get('mandatory_unit') == "Weight": 
+					request_data['doctype_data']['unit_weight'] = self.metrc_unit_value
+					request_data['doctype_data']['unit_weight_uom'] = self.metrc_unit_uom
+
+			# create am item on the bloomtrace
+			bloomtrace_response = requests.post('https://bl2qu9obqb.execute-api.ap-south-1.amazonaws.com/dev/doctype/create-item', json=request_data)
```
Comment (rvpasari) [inline erpnext/stock/doctype/item/item.py:952]: same, url needs to come from bloomtrace settings

Comment (rvpasari) [reply erpnext/stock/doctype/item/item.py:952]: raise a PR for the settings and mark this PR as dependent on that

**Review by rvpasari — CHANGES_REQUESTED** [review]

File: erpnext/stock/doctype/item/item.py (hunk 7 above)
Comment (harshmule-epi) [reply erpnext/stock/doctype/item/item.py:952]: yeah, raising a PR for bloomtrace settings

File: erpnext/stock/doctype/item/item.py (hunk 6 above)
Comment (harshmule-epi) [reply erpnext/stock/doctype/item/item.py]: only this account can get the user key from admin portal or else it's failing at metrc post, also we cannot put the owner/modified by in the bloomtrace settings, right??

**Pull request closed without merging**